src/file-manager/fm-icon-container.c

```diff
@@ -295,6 +296,72 @@ fm_icon_container_get_icon_text_attribute_names (CajaIconContainer *container,
     return attributes;
 }
 
+/**
+ * Parse a .git directory/file to determine repository branch name
+ **/
+static char *
+get_git_branch (const char *git_path) {
+    gchar *resolved_git_dir = NULL;
+    gchar *head_content = NULL;
+    gchar *branch_name = NULL;
+    gchar *head_path = NULL;
+    gsize len = 0;
+
+    if (g_file_test (git_path, G_FILE_TEST_IS_DIR))
+    {
+        resolved_git_dir = g_strdup (git_path);
+    } else if (g_file_test (git_path, G_FILE_TEST_IS_REGULAR))
+    {
+        /* It's a file, read it and resolve the real git dir */
+        gchar *contents = NULL;
+        if (g_file_get_contents (git_path, &contents, NULL, NULL) && contents)
+        {
+            if (g_str_has_prefix (contents, "gitdir: "))
+            {
+                gchar *relative = g_strstrip (contents + 8);
```

Comment:
Not tested, but avoids a ‘magic number’ from the outset.
```suggestion
            const char *git_dir_prefix = "gitdir: ";
            if (g_str_has_prefix (contents, git_dir_prefix))
            {
                gsize git_dir_prefix_len = g_ref_string_length (git_dir_prefix);
                gchar *relative = g_strstrip (contents + git_dir_prefix_len);
```

Reply:
Yes, [it does](https://godbolt.org/z/37cEzhfnn) if you use `strlen` and enable optimizations.  
But I spotted a problem with my suggestion while checking what the implementation of `g_ref_string_length` is to check if it'd also be optimized away on constant strings. It expects a `GRefStringImpl` struct but the constant isn't a GLib string. Thus it should be `size_t git_dir_prefix_len = strlen (git_dir_prefix);` and will hopefully work :smile:

Reply:
Having now had a fresh look at the function, i believe this minor refactor is somewhat cleaner:

/*
 The next step would be to handle submodules within submodules, but i have not created
 (or even know if such scenarios are possible) any test cases to work with, so for now,
 this shall have to be good enough.
  */
static char *
get_git_branch (const char *git_path) {
    gchar *head_content = NULL;
    gchar *branch_name = NULL;
    gchar *head_path = NULL;

    if (g_file_test (git_path, G_FILE_TEST_IS_REGULAR))
    {
        /* It's a file, thus we are probably dealing with a submodule, so read it to resolve the real git dir */
        gchar *contents = NULL;
        if (g_file_get_contents (git_path, &contents, NULL, NULL) && contents)
        {
            const char *git_dir_prefix = "gitdir: ";
            if (g_str_has_prefix (contents, git_dir_prefix))
            {
                size_t git_dir_prefix_len = strlen (git_dir_prefix);
                gchar *relative = g_strstrip (contents + git_dir_prefix_len);
                gchar *base = g_path_get_dirname (git_path);
                g_free (git_path);
                git_path = g_build_filename (base, relative, NULL);
                g_free (base);
            }
            g_free (contents);
        }
    }
    
    /* Extract the current git branch name of the repository from file HEAD within .git directory */
    if (g_file_test (git_path, G_FILE_TEST_IS_DIR))
    {
        head_path = g_build_filename (git_path, "HEAD", NULL);
        if (g_file_get_contents (head_path, &head_content, NULL, NULL))
        {
            g_strstrip (head_content);
            if (g_str_has_prefix (head_content, "ref: "))
            {
                gchar **parts = g_strsplit (head_content, "/", -1);
                if (parts != NULL)
                {
                    branch_name = g_strdup (parts[g_strv_length(parts) - 1]);
                    g_strchomp (branch_name);
                    g_strfreev (parts);
                }
            }
            else
            {
                /* Repository is in a detached HEAD state */
                branch_name = g_strdup_printf (_("detached: %.7s"), head_content);
            }
        }
    }

    g_free (head_content);
    g_free (head_path);

    /* branch_name is a newly allocated string, so the caller is responsible for freeing it */
    return branch_name;
}

Reply:
btw, is the use of recursion banned from this code-base?